src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs

```diff
@@ -0,0 +1,1864 @@
+                    throw new ArgumentException(SR.Argument_IncompatibleArrayType, nameof(array));
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        int IList.Add(object? value)
+        {
+            if (value is not KeyValuePair<TKey, TValue> pair)
+            {
+                throw new ArgumentException(SR.Format(SR.Arg_WrongType, value, typeof(KeyValuePair<TKey, TValue>)), nameof(value));
+            }
+
+            Add(pair.Key, pair.Value);
+            return Count - 1;
+        }
+
+        /// <inheritdoc/>
+        bool IList.Contains(object? value) =>
+            value is KeyValuePair<TKey, TValue> pair &&
+            TryGetValue(pair.Key, out TValue? v) &&
+            EqualityComparer<TValue>.Default.Equals(v, pair.Value);
+
+        /// <inheritdoc/>
+        int IList.IndexOf(object? value)
+        {
+            if (value is KeyValuePair<TKey, TValue> pair)
+            {
+                return ((IList<KeyValuePair<TKey, TValue>>)this).IndexOf(pair);
+            }
+
+            return -1;
+        }
+
+        /// <inheritdoc/>
+        void IList.Insert(int index, object? value)
+        {
+            if (value is not KeyValuePair<TKey, TValue> pair)
+            {
+                throw new ArgumentException(SR.Format(SR.Arg_WrongType, value, typeof(KeyValuePair<TKey, TValue>)), nameof(value));
+            }
+
+            Insert(index, pair.Key, pair.Value);
+        }
+
+        /// <inheritdoc/>
+        void IList.Remove(object? value)
+        {
+            if (value is KeyValuePair<TKey, TValue> pair)
+            {
+                ((ICollection<KeyValuePair<TKey, TValue>>)this).Remove(pair);
+            }
+        }
+
+        /// <summary>Represents a key/value pair in the dictionary.</summary>
+        private struct Entry
+        {
+            /// <summary>The index of the next entry in the chain, or -1 if this is the last entry in the chain.</summary>
+            public int Next;
+            /// <summary>Cached hash code of <see cref="Key"/>.</summary>
+            public uint HashCode;
+            /// <summary>The key.</summary>
+            public TKey Key;
+            /// <summary>The value associated with <see cref="Key"/>.</summary>
+            public TValue Value;
+        }
+
+        /// <summary>Enumerates the elements of a <see cref="OrderedDictionary{TKey, TValue}"/>.</summary>
+        [StructLayout(LayoutKind.Auto)]
+        public struct Enumerator : IEnumerator<KeyValuePair<TKey, TValue>>, IDictionaryEnumerator
+        {
+            /// <summary>The dictionary being enumerated.</summary>
+            private readonly OrderedDictionary<TKey, TValue> _dictionary;
+            /// <summary>A snapshot of the dictionary's version when enumeration began.</summary>
+            private readonly int _version;
+            /// <summary>Whether Current should be a DictionaryEntry.</summary>
+            private readonly bool _useDictionaryEntry;
```

Comment:
Is this hybrid mode used  by the dictionary enumerator as well?

Reply:
Yup